File: examples/rust/player/agg/src/state.rs
//! Player aggregate state.
//!
//! DOC: This file is referenced in docs/docs/examples/aggregates.mdx
//!      Update documentation when making changes to StateRouter patterns.

use std::collections::HashMap;
use std::sync::LazyLock;

use angzarr_client::proto::examples::{
    FundsDeposited, FundsReleased, FundsReserved, FundsTransferred, FundsWithdrawn,
    PlayerRegistered, PlayerState as ProtoPlayerState, PlayerType,
};
use angzarr_client::proto::EventBook;
use angzarr_client::StateRouter;
use angzarr_client::UnpackAny;

/// Player aggregate state rebuilt from events.
#[derive(Debug, Default, Clone)]
pub struct PlayerState {
    pub player_id: String,
    pub display_name: String,
    pub email: String,
    pub player_type: PlayerType,
    pub ai_model_id: String,
    pub bankroll: i64,
    pub reserved_funds: i64,
    pub table_reservations: HashMap<String, i64>, // table_root_hex -> amount
    pub status: String,
}

impl PlayerState {
    /// Check if the player exists.
    pub fn exists(&self) -> bool {
        !self.player_id.is_empty()
    }

    /// Get available balance (bankroll - reserved).
    pub fn available_balance(&self) -> i64 {
        self.bankroll - self.reserved_funds
    }

    /// Check if this is an AI player.
    pub fn is_ai(&self) -> bool {
        self.player_type == PlayerType::Ai
    }
}

// Event applier functions for StateRouter

// docs:start:state_router
fn apply_registered(state: &mut PlayerState, event: PlayerRegistered) {
    state.player_id = format!("player_{}", event.email);
    state.display_name = event.display_name;
    state.email = event.email;
    state.player_type = PlayerType::try_from(event.player_type).unwrap_or_default();
    state.ai_model_id = event.ai_model_id;
    state.status = "active".to_string();
    state.bankroll = 0;
    state.reserved_funds = 0;
}

fn apply_deposited(state: &mut PlayerState, event: FundsDeposited) {
    if let Some(balance) = event.new_balance {
        state.bankroll = balance.amount;
    }
}

fn apply_withdrawn(state: &mut PlayerState, event: FundsWithdrawn) {
    if let Some(balance) = event.new_balance {
        state.bankroll = balance.amount;
    }
}

fn apply_reserved(state: &mut PlayerState, event: FundsReserved) {
    if let Some(balance) = event.new_reserved_balance {
        state.reserved_funds = balance.amount;
    }
    if let (Some(amount), table_root) = (event.amount, event.table_root) {
        let table_key = hex::encode(&table_root);
        state.table_reservations.insert(table_key, amount.amount);
    }
}

fn apply_released(state: &mut PlayerState, event: FundsReleased) {
    if let Some(balance) = event.new_reserved_balance {
        state.reserved_funds = balance.amount;
    }
    let table_key = hex::encode(&event.table_root);
    state.table_reservations.remove(&table_key);
}

fn apply_transferred(state: &mut PlayerState, event: FundsTransferred) {
    if let Some(balance) = event.new_balance {
        state.bankroll = balance.amount;
    }
}

/// StateRouter for fluent state reconstruction.
static STATE_ROUTER: LazyLock<StateRouter<PlayerState>> = LazyLock::new(|| {
    StateRouter::new()
        .on::<PlayerRegistered>("PlayerRegistered", apply_registered)
        .on::<FundsDeposited>("FundsDeposited", apply_deposited)
        .on::<FundsWithdrawn>("FundsWithdrawn", apply_withdrawn)
        .on::<FundsReserved>("FundsReserved", apply_reserved)
        .on::<FundsReleased>("FundsReleased", apply_released)
        .on::<FundsTransferred>("FundsTransferred", apply_transferred)
});
// docs:end:state_router

/// Rebuild player state from event history.
pub fn rebuild_state(event_book: &EventBook) -> PlayerState {
    // Start from snapshot if available
    if let Some(snapshot) = &event_book.snapshot {
        if let Some(snapshot_any) = &snapshot.state {
            if let Ok(proto_state) = snapshot_any.unpack::<ProtoPlayerState>() {
                let mut state = apply_snapshot(&proto_state);
                // Apply events since snapshot
                for page in &event_book.pages {
                    if let Some(event) = &page.event {
                        STATE_ROUTER.apply_single(&mut state, event);
                    }
                }
                return state;
            }
        }
    }

    STATE_ROUTER.with_event_book(event_book)
}

fn apply_snapshot(snapshot: &ProtoPlayerState) -> PlayerState {
    let bankroll = snapshot.bankroll.as_ref().map(|c| c.amount).unwrap_or(0);
    let reserved_funds = snapshot
        .reserved_funds
        .as_ref()
        .map(|c| c.amount)
        .unwrap_or(0);

    PlayerState {
        player_id: snapshot.player_id.clone(),
        display_name: snapshot.display_name.clone(),
        email: snapshot.email.clone(),
        player_type: PlayerType::try_from(snapshot.player_type).unwrap_or_default(),
        ai_model_id: snapshot.ai_model_id.clone(),
        bankroll,
        reserved_funds,
        table_reservations: snapshot.table_reservations.clone(),
        status: snapshot.status.clone(),
    }
}

File: examples/python/player/agg/state.py
"""Player state - functional pattern using StateRouter.

This file defines the player state and event appliers as pure functions.
Contrasts with the OO pattern in player/agg/handlers/player.py which
uses decorators on class methods.
"""

from dataclasses import dataclass, field

from angzarr_client.proto.examples import player_pb2 as player


@dataclass
class PlayerState:
    """Player aggregate state."""

    player_id: str = ""
    display_name: str = ""
    email: str = ""
    player_type: int = 0
    ai_model_id: str = ""
    bankroll: int = 0
    reserved_funds: int = 0
    table_reservations: dict = field(default_factory=dict)
    status: str = ""

    @property
    def exists(self) -> bool:
        return bool(self.player_id)

    @property
    def available_balance(self) -> int:
        return self.bankroll - self.reserved_funds


# --- Event appliers (pure functions) ---


# docs:start:state_router
def apply_registered(state: PlayerState, event: player.PlayerRegistered) -> None:
    """Apply PlayerRegistered event to state."""
    state.player_id = f"player_{event.email}"
    state.display_name = event.display_name
    state.email = event.email
    state.player_type = event.player_type
    state.ai_model_id = event.ai_model_id
    state.status = "active"
    state.bankroll = 0
    state.reserved_funds = 0


def apply_deposited(state: PlayerState, event: player.FundsDeposited) -> None:
    """Apply FundsDeposited event to state."""
    if event.new_balance:
        state.bankroll = event.new_balance.amount


def apply_withdrawn(state: PlayerState, event: player.FundsWithdrawn) -> None:
    """Apply FundsWithdrawn event to state."""
    if event.new_balance:
        state.bankroll = event.new_balance.amount


def apply_reserved(state: PlayerState, event: player.FundsReserved) -> None:
    """Apply FundsReserved event to state."""
    if event.new_reserved_balance:
        state.reserved_funds = event.new_reserved_balance.amount
    table_key = event.table_root.hex()
    if event.amount:
        state.table_reservations[table_key] = event.amount.amount


def apply_released(state: PlayerState, event: player.FundsReleased) -> None:
    """Apply FundsReleased event to state."""
    if event.new_reserved_balance:
        state.reserved_funds = event.new_reserved_balance.amount
    table_key = event.table_root.hex()
    state.table_reservations.pop(table_key, None)


def apply_transferred(state: PlayerState, event: player.FundsTransferred) -> None:
    """Apply FundsTransferred event to state."""
    if event.new_balance:
        state.bankroll = event.new_balance.amount


# docs:end:state_router


def build_state(state: PlayerState, events: list) -> PlayerState:
    """Build state from a list of Any-wrapped events.

    Args:
        state: Initial state to mutate.
        events: List of Any-wrapped protobuf events.

    Returns:
        The mutated state.
    """
    from google.protobuf.any_pb2 import Any as AnyProto

    _appliers = {
        "examples.PlayerRegistered": (player.PlayerRegistered, apply_registered),
        "examples.FundsDeposited": (player.FundsDeposited, apply_deposited),
        "examples.FundsWithdrawn": (player.FundsWithdrawn, apply_withdrawn),
        "examples.FundsReserved": (player.FundsReserved, apply_reserved),
        "examples.FundsReleased": (player.FundsReleased, apply_released),
        "examples.FundsTransferred": (player.FundsTransferred, apply_transferred),
    }

    for event_any in events:
        if not isinstance(event_any, AnyProto):
            continue
        # Extract type name from type_url (e.g., "type.googleapis.com/examples.PlayerRegistered")
        type_name = event_any.type_url.split("/")[-1]
        if type_name in _appliers:
            proto_cls, applier = _appliers[type_name]
            event = proto_cls()
            event_any.Unpack(event)
            applier(state, event)

    return state
